fili-core/src/main/java/com/yahoo/bard/webservice/table/MetricUnionCompositeTable.java

```diff
@@ -0,0 +1,90 @@
+import com.yahoo.bard.webservice.data.time.ZonedTimeGrain;
+import com.yahoo.bard.webservice.table.availability.MetricUnionAvailability;
+import com.yahoo.bard.webservice.table.resolver.GranularityComparator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An implementation of <tt>BasePhysicalTable</tt> backed by metric union availability.
+ */
+public class MetricUnionCompositeTable extends BasePhysicalTable {
+    /**
+     * Constructor.
+     *
+     * @param name  Name of the physical table as TableName, also used as fact table name
+     * @param columns  The columns for this table
+     * @param physicalTables  A set of <tt>PhysicalTable</tt>s
+     * @param logicalToPhysicalColumnNames  Mappings from logical to physical names
+     */
+    public MetricUnionCompositeTable(
+            @NotNull TableName name,
+            @NotNull Set<Column> columns,
+            @NotNull Set<PhysicalTable> physicalTables,
+            @NotNull Map<String, String> logicalToPhysicalColumnNames
+    ) {
+        super(
+                name,
+                getCoarsestTimeGrain(physicalTables),
+                columns,
+                logicalToPhysicalColumnNames,
+                new MetricUnionAvailability(physicalTables, columns)
+        );
+    }
+
+    /**
+     * Returns the coarsest <tt>ZonedTimeGrain</tt> that satisfies all tables.
+     * <p>
+     * If the set of <tt>PhysicalTables</tt>'s is empty or the coarsest <tt>ZonedTimeGrain</tt> is not
+     * compatible with any of the <tt>PhysicalTables</tt>s, throw <tt>IllegalArgumentException</tt>.
+     *
+     * @param physicalTables  A set of <tt>PhysicalTable</tt>s among which the coarsest <tt>ZonedTimeGrain</tt>
+     * is to be returned.
+     *
+     * @return the coarsest <tt>ZonedTimeGrain</tt> among a set of <tt>PhysicalTables</tt>s
+     */
+    private static ZonedTimeGrain getCoarsestTimeGrain(Set<PhysicalTable> physicalTables) {
+        if (physicalTables.isEmpty()) {
+            throw new IllegalArgumentException("At least 1 physical table needs to be provided");
+        }
+
+        GranularityComparator granularityComparator = new GranularityComparator();
+
+        // sort tables by <tt>ZonedTimeGrain</tt> in increasing order
+        List<PhysicalTable> sortedTables = physicalTables.stream()
+                .sorted((table1, table2) -> granularityComparator.compare(table1, table2))
+                .collect(Collectors.toList());
+
+        // check to see if all <tt>ZonedTimeGrain</tt>'s is compatible with the coarsest <tt>ZonedTimeGrain</tt>
+        ZonedTimeGrain coarsestTimeGrain = sortedTables.get(0).getSchema().getTimeGrain();
+        List<PhysicalTable> incompatibles = sortedTables.stream()
+                .filter(table -> !table.getSchema().getTimeGrain().satisfiedBy(coarsestTimeGrain))
```

Comment:
Rather than embedding this call string into the filter, it's generally better to pivot it to be declarative via `map` calls if possible. Once that's done, I think this helper method may be useful:

```java
private static Stream<ZonedTimeGrain> getSortedZonedTimeGrainStream(List<PhysicalTable> sortedTables) {
    return sortedTables.stream()
            .map(PhysicalTable::getSchema)
            .map(PhysicalTableSchema::getTimeGrain);
}
```

Especially if coupled with a more generic condition, rather than accumulating things for which we're not sure of the true error:

```java
if (getSortedZonedTimeGrainStream(sortedTables).anyMatch(ztg -> !ztg.satisfiedBy(coarsestTimeGrain))) {
```